### PR TITLE
Better delete

### PR DESCRIPTION
**Pull request by jreidinger**
PBI: https://trello.com/c/C0QFH0Su/1013-8-new-storage-new-expert-partitioner-4-write-operations-for-hard-disks

how it looks like:

snimek obrazovky_2017-07-03_12-58-45

### CODE REVIEW

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 1]
```diff
@@ -28,18 +28,22 @@ def label
...
+        if device.nil?
+          Yast::Popup.Error(_("No device selected"))
+          return nil
+        end
 
-        return nil unless ret
+        return nil unless confirm(device)
 
         partition_table = device.partition_table
         if device.is?(:disk)
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Remember that, so far, this returns false for DASD. Intended?

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 2]
```diff
@@ -28,18 +28,22 @@ def label
...
-          partition_table.partitions.each { |p| partition_table.delete_partition(p) }
+          partitions = partition_table.partitions.reject { |p| p.type.is?(:logical) }
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Is deleting the partitions enough? Don't we want to also wipe the partition table itself?

Which makes me wonder.... maybe it's useful to move this logic to partition so we have something like `PartitionTable#delete_partitions` and we don't have to always remember to filter out the logical ones?

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 3]
```diff
@@ -48,7 +52,42 @@ def handle
 
     private
 
+      def confirm(device)
+        names = children_names(device)
+
+        if names.empty?
+          # TODO: check for children that will die and if there are, use confirm_recursive_delete
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: This comment does not longer apply

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 4]
```diff
@@ -48,7 +52,42 @@ def handle
...
+          Yast::Popup.YesNo(
+            # TRANSLATORS %s is device to be deleted
+            format(_("Really delete %s?"), device.name)
+          )
+        else
+          confirm_recursive_delete(
+            names,
+            _("Confirm Deleting of All Partitions"),
+            format(_("The %{type} \"%{name}\" contains at least one another device.\n" \
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: This needs some note for the translators about the placeholders

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 5]
```diff
@@ -48,7 +52,42 @@ def handle
...
+              "If you proceed, the following devices will be deleted:"),
+              name: device.name,
+              type: device.is?(:disk) ? _("disk") : _("partition")),
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Once again, remember this doesn't include DASD. Maybe you want to consider to add a `:disk_device` (or similar) handler for `Disk#is?` and `Dasd#is?` so you can ask `device.is?(:disk_device)`. Kind of symmetric to [`Devicegraph#disk_devices](https://github.com/yast/yast-storage-ng/blob/master/src/lib/y2storage/devicegraph.rb#L155)

Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Or, of course, add for `is?(:partition)` instead :)

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 6]
```diff
@@ -48,7 +52,42 @@ def handle
...
+        end
+      end
+
+      def children_names(device)
+        device.descendants.map do |dev|
+          if dev.respond_to?(:name)
+            dev.name
+          elsif dev.is?(:lvm_vg)
+            # TODO: discuss to have it in storage-ng
+            "/dev/#{dev.vg_name}"
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Please, do so (adding it to storage-ng)

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb (hunk 1 above)
Comment (jreidinger) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: yes, so far no support for s390 in partitioner, also we should support here other devices like LVM, or RAD which can be deleted.

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb (hunk 2 above)
Comment (jreidinger) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: it is fine for me. Regarding wiping our partition table, I have no clue. old libstorage just mark partition to delete.

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb (hunk 5 above)
Comment (jreidinger) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: in future I will ahve to extend it by lvm or raid, so it is just first step...and for sure dasd will need also improvement

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb (hunk 6 above)
Comment (jreidinger) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: so you are fine with adding `name` to lvm_vg, right?

Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Right

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb (hunk 2 above)
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: Let's keep the behavior then. I just wanted to be sure.

Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: I checked and found there is a separate "Expert -> Create New Partition Table" option to mess with whole partition tables. Curious how the UI have evolved over time :-)

File: src/lib/y2partitioner/widgets/delete_disk_partition_button.rb [hunk 7]
```diff
@@ -48,7 +49,37 @@ def handle
...
+              type: device.is?(:disk) ? _("disk") : _("partition")),
+            format(_("Really delete all devices on \"%s\"?"), device.name)
+          )
+        end
+      end
+
+      def children_names(device)
+        device.descendants.map do |dev|
+          dev.name if dev.respond_to?(:name)
+        end.compact
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: I'm fine with this. But somebody I know called J.R. would ask you to use `each_with_object` here.

Comment (jreidinger) [inline src/lib/y2partitioner/widgets/delete_disk_partition_button.rb]: :smile:

**Pull request merged**